### PR TITLE
add endpoint to check if email is already registered

### LINKED ISSUE #67: Endpoint to check if email is used for another account
Used to check if an email is available during registration.
image

### PR DESCRIPTION
**Pull request by fremartini**
closes #67

### CODE REVIEW

File: coffeecard/CoffeeCard.WebApi/Controllers/v2/AccountController.cs [hunk 1]
```diff
@@ -46,5 +47,24 @@ public async Task<ActionResult> Delete()
...
             return NoContent();
         }
+        
+        /// <summary>
+        /// Check if a given email is in use
+        /// </summary>
+        /// <returns>Account information</returns>
+        /// <response code="200">Successful request</response>
+        /// <response code="401">Invalid credentials</response>
+        [HttpPost]
```
Comment (jonasanker) [inline coffeecard/CoffeeCard.WebApi/Controllers/v2/AccountController.cs:57]: The endpoint should be anonymous
```csharp
[AllowAnonymous]
```

File: coffeecard/CoffeeCard.WebApi/Controllers/v2/AccountController.cs [hunk 2]
```diff
@@ -46,5 +47,24 @@ public async Task<ActionResult> Delete()
...
+        [ProducesResponseType(typeof(EmailInUseDTO), StatusCodes.Status200OK)]
+        [ProducesResponseType(typeof(void), StatusCodes.Status401Unauthorized)]
+        [Route("emailinuse")]
+        public async Task<ActionResult<EmailInUseDTO>> EmailInUse([FromBody] string email)
```
Comment (jonasanker) [inline coffeecard/CoffeeCard.WebApi/Controllers/v2/AccountController.cs:61]: - Input should be an object
- Document parameter in above code comments

File: coffeecard/CoffeeCard.Models/DataTransferObjects/User/EmailInUseDTO.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+using System.ComponentModel.DataAnnotations;
+
+namespace CoffeeCard.Models.DataTransferObjects.User
+{
+    public class EmailInUseDTO
```
Comment (jonasanker) [inline coffeecard/CoffeeCard.Models/DataTransferObjects/User/EmailInUseDTO.cs:5]: 1) Shold be added to the v2 subfolder
2) In Api V2, we tend to not the the DTO suffix but instead give a more direct name, e.g. `EmailInUseResponse`

File: coffeecard/CoffeeCard.WebApi/Controllers/v2/AccountController.cs [hunk 4]
```diff
@@ -46,5 +47,24 @@ public async Task<ActionResult> Delete()
...
+        /// <summary>
+        /// Check if a given email is in use
+        /// </summary>
+        /// <returns>Account information</returns>
+        /// <response code="200">Successful request</response>
+        /// <response code="401">Invalid credentials</response>
+        [HttpPost]
+        [ProducesResponseType(typeof(EmailInUseDTO), StatusCodes.Status200OK)]
+        [ProducesResponseType(typeof(void), StatusCodes.Status401Unauthorized)]
+        [Route("emailinuse")]
```
Comment (jonasanker) [inline coffeecard/CoffeeCard.WebApi/Controllers/v2/AccountController.cs:60]: In Rest, we usually name the action with a verb. This is understood as we _apply_ an action on the REST collection. With that in mind, I would like to find another name than _emailInUse_. 

Suggestions
- `/check-email`
- `/check-duplicate-email`

**Review by jonasanker — CHANGES_REQUESTED** [review]
Changes Requested

File: coffeecard/CoffeeCard.Models/DataTransferObjects/v2/User/EmailInUseRequest.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
+using System.ComponentModel.DataAnnotations;
+
+namespace CoffeeCard.Models.DataTransferObjects.v2.User
+{
+    /// <summary>
+    /// An email that should be checked if it already exists
+    /// </summary>
+    /// <example>
+    /// {
+    ///     "email": "johndoe@mail.com",
```
Comment (jonasanker) [inline coffeecard/CoffeeCard.Models/DataTransferObjects/v2/User/EmailInUseRequest.cs:10]: No comma here

File: coffeecard/CoffeeCard.Models/DataTransferObjects/v2/User/EmailInUseResponse.cs [hunk 6]
```diff
@@ -0,0 +1,21 @@
+using System.ComponentModel.DataAnnotations;
+
+namespace CoffeeCard.Models.DataTransferObjects.v2.User
+{
+    /// <summary>
+    /// Contains information about an email being in use
+    /// </summary>
+    /// <example>
+    /// {
+    ///     "InUse": true,
```
Comment (jonasanker) [inline coffeecard/CoffeeCard.Models/DataTransferObjects/v2/User/EmailInUseResponse.cs:10]: No comma

File: coffeecard/CoffeeCard.WebApi/Controllers/v2/AccountController.cs [hunk 7]
```diff
@@ -46,5 +48,25 @@ public async Task<ActionResult> Delete()
...
+        /// Check if a given email is in use
+        /// </summary>
+        /// <returns>Account information</returns>
+        /// <response code="200">Successful request</response>
+        /// <response code="401">Invalid credentials</response>
+        [HttpPost]
+        [ProducesResponseType(typeof(EmailInUseResponse), StatusCodes.Status200OK)]
+        [ProducesResponseType(typeof(void), StatusCodes.Status401Unauthorized)]
+        [AllowAnonymous]
+        [Route("check-email-in-use")]
```
Comment (jonasanker) [inline coffeecard/CoffeeCard.WebApi/Controllers/v2/AccountController.cs:62]: I think _email in use_ is a bit clumsy name compared to e.g. email exists

File: coffeecard/CoffeeCard.WebApi/Controllers/v2/AccountController.cs [hunk 8]
```diff
@@ -46,5 +48,25 @@ public async Task<ActionResult> Delete()
...
+        public async Task<ActionResult<EmailInUseResponse>> EmailInUse([FromBody] EmailInUseRequest request)
```
Comment (jonasanker) [inline coffeecard/CoffeeCard.WebApi/Controllers/v2/AccountController.cs:63]: We should add code documentation for this parameter

File: coffeecard/CoffeeCard.Models/DataTransferObjects/v2/User/EmailInUseResponse.cs [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    /// }
+    /// </example>
+    public class EmailInUseResponse
+    {
+        /// <summary>
+        /// Contains information about an email being in use
+        /// </summary>
+        [Required]
+        public bool InUse { get; set; }
```
Comment (jonasanker) [inline coffeecard/CoffeeCard.Models/DataTransferObjects/v2/User/EmailInUseResponse.cs:19]: Suggestion: Email exists

**Review by jonasanker — CHANGES_REQUESTED** [review]

File: coffeecard/CoffeeCard.WebApi/Controllers/v2/AccountController.cs [hunk 10]
```diff
@@ -46,5 +48,26 @@ public async Task<ActionResult> Delete()
...
+        /// <param name="request">The email that should be checked</param>
+        /// <returns>Account information</returns>
+        /// <response code="200">Successful request</response>
+        /// <response code="401">Invalid credentials</response>
+        [HttpPost]
+        [ProducesResponseType(typeof(EmailInUseResponse), StatusCodes.Status200OK)]
+        [ProducesResponseType(typeof(void), StatusCodes.Status401Unauthorized)]
+        [AllowAnonymous]
+        [Route("email-exists")]
+        public async Task<ActionResult<EmailInUseResponse>> EmailInUse([FromBody] EmailInUseRequest request)
```
Comment (jonasanker) [inline coffeecard/CoffeeCard.WebApi/Controllers/v2/AccountController.cs:64]: With the update route value, the method name should be updated to align. Same goes with request and response class name 😉

**Review by jonasanker — CHANGES_REQUESTED** [review]

**Review by jonasanker — CHANGES_REQUESTED** [review]

**Review by jonasanker — CHANGES_REQUESTED** [review]

File: coffeecard/CoffeeCard.Models/DataTransferObjects/v2/User/EmailExistsResponse.cs [hunk 11]
```diff
@@ -0,0 +1,21 @@
+using System.ComponentModel.DataAnnotations;
+
+namespace CoffeeCard.Models.DataTransferObjects.v2.User
+{
+    /// <summary>
+    /// Contains information about an email being in use
+    /// </summary>
+    /// <example>
+    /// {
+    ///     "EmailExists": true
```
[the remark this answers is not in the document]
Comment (fremartini) [reply coffeecard/CoffeeCard.Models/DataTransferObjects/v2/User/EmailExistsResponse.cs:10]: it already is? EmailExists?

[the remark this answers is not in the document]
Comment (jonasanker) [reply coffeecard/CoffeeCard.Models/DataTransferObjects/v2/User/EmailExistsResponse.cs:10]: It's not in the example

**Review by jonasanker — APPROVED** [review]

**Pull request merged**